core/src/scene/stops.cpp

```diff
@@ -11,33 +14,109 @@
+            color = (color << 8) + colorNode[2].as<float>() * 255.;
+            color = (color << 8) + colorNode[1].as<float>() * 255.;
+            color = (color << 8) + colorNode[0].as<float>() * 255.;
+        }
+        stops.frames.emplace_back(key, color);
+    }
+    return stops;
+}
+
+double widthMeterToPixel(float _zoom, double _tileSize, double _width) {
+    // pixel per meter at z == 0
+    double meterRes = _tileSize / (2.0 * MapProjection::HALF_CIRCUMFERENCE);
+    // pixel per meter at zoom
+    meterRes *= exp2(_zoom);
+
+    return _width * meterRes;
+}
+
+auto Stops::Width(const YAML::Node& _node, const MapProjection& _projection) -> Stops {
+    Stops stops;
+    if (!_node.IsSequence()) { return stops; }
+
+    double tileSize = _projection.TileSize();
 
-    if (!_node.IsSequence()) { return; }
+    bool lastIsMeter;
+    float lastKey = 0;
+    float lastMeter = 0;
 
     for (const auto& frameNode : _node) {
         if (!frameNode.IsSequence() || frameNode.size() != 2) { continue; }
         float key = frameNode[0].as<float>();
-        if (_isColor) {
-            // parse color from sequence or string
-            uint32_t color = 0;
-            YAML::Node colorNode = frameNode[1];
-            if (colorNode.IsScalar()) {
-                color = CSSColorParser::parse(colorNode.as<std::string>()).getInt();
-            } else if (colorNode.IsSequence() && colorNode.size() >= 3) {
-                float alpha = colorNode.size() > 3 ? colorNode[3].as<float>() : 1.f;
-                color = alpha * 255.;
-                color = (color << 8) + colorNode[2].as<float>() * 255.;
-                color = (color << 8) + colorNode[1].as<float>() * 255.;
-                color = (color << 8) + colorNode[0].as<float>() * 255.;
+
+        if (lastKey > key) {
+            logMsg("Invalid stop order: key %f > %f\n", lastKey, key);
+            continue;
+        }
+        lastKey = key;
+
+        StyleParam::ValueUnitPair width;
+        width.unit = Unit::meter;
+        size_t start = 0;
+
+        if (StyleParam::parseValueUnitPair(frameNode[1].Scalar(), start, width)){
+            if (width.unit == Unit::meter) {
+                float w = widthMeterToPixel(key, tileSize, width.value);
+                stops.frames.emplace_back(key, w);
+
+                lastIsMeter = true;
+                lastMeter = width.value;
+
+            } else {
+                stops.frames.emplace_back(key, width.value);
+                lastIsMeter = false;
+
             }
-            frames.emplace_back(key, color);
         } else {
-            // parse distance from string
-            float value = frameNode[1].as<float>();
-            frames.emplace_back(key, value);
+            logMsg("could not parse node %s\n", Dump(frameNode[1]).c_str());
         }
     }
+    // Append stop at max-zoom to continue scaling after the last stop
+    // TODO: define MAX_ZOOM == 24
+    if (lastIsMeter && lastKey < 24) {
```

Comment:
Only has to be not smaller
On Oct 6, 2015 9:44 AM, "Varun" <notifications@github.com> wrote:

> In core/src/scene/stops.cpp
> <https://github.com/tangrams/tangram-es/pull/316#discussion_r41264805>:
>
> >          }
> >      }
> > +    // Append stop at max-zoom to continue scaling after the last stop
> > +    // TODO: define MAX_ZOOM == 24
> > +    if (lastIsMeter && lastKey < 24) {
>
> Should max zoom be same as the view max zoom, here
> <https://github.com/tangrams/tangram-es/blob/master/core/src/view/view.h#L122>
> ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tangrams/tangram-es/pull/316/files#r41264805>.
>